Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,658 @@
+        if (err)
+            LogMsg(kLogError, "Failed to shutdown socket: {}", err.message());
+        
+        // Add the socket to the close list in sorted order by close time;
+        // if this socket is the first on the list then start the timer
+        bool startTimer;
+        {
+            hsLockGuard(s_socketCrit);
+            s_socketList.push_back(conn);
+            startTimer = s_socketList.front() == conn;
+        }
+        
+        // If this is the first item queued in the socket list then start timer.
+        // This operation should be safe to perform outside the critical section
+        // because s_socketTimer should not be deleted before application shutdown
+        if (startTimer)
+            AsyncTimerUpdate(s_socketTimer, kCloseTimeoutMs);
+    }
+}
+
+static bool SocketQueueAsyncWrite(AsyncSocket conn, const void* data, unsigned bytes)
+{
+    hsLockGuard(s_connectCrit);
+    
+    // check for data backlog
+    WriteOperation * firstQueuedWrite = conn->fWriteOps.front();
+    if  (firstQueuedWrite) {
+        unsigned currTimeMs = TimeGetMs();
+        if (((long) (currTimeMs - firstQueuedWrite->queueTimeMs) >= (long) kBacklogFailMs)
+        &&  ((long) (currTimeMs - conn->initTimeMs) >= (long) kBacklogInitMs)
+        ) {
+            PerfAddCounter(kAsyncPerfSocketDisconnectBacklog, 1);
+
+            if (conn->fConnectionType) {
+                LogMsg(
+                    kLogPerf,
+                    "Backlog, c:{} q:{}, i:{}",
+                    conn->fConnectionType,
+                    currTimeMs - firstQueuedWrite->queueTimeMs,
+                    currTimeMs - conn->initTimeMs
+                );
+            }
+            AsyncSocketDisconnect((AsyncSocket) conn, true);
+            return false;
+        }
+    }
+
+    // If the last buffer still has space available then add data to it
+    if (!conn->fWriteOps.empty()) {
+        WriteOperation* op = conn->fWriteOps.back();
+        unsigned bytesLeft = std::min(op->fAllocSize - op->fNotify.bytes, bytes);
+        if (bytesLeft) {
+            PerfAddCounter(kAsyncPerfSocketBytesWaitQueued, bytesLeft);
+            memcpy(op->fNotify.buffer + op->fNotify.bytes, data, bytesLeft);
+            op->fNotify.bytes += bytesLeft;
+            data = (const uint8_t *)data + bytesLeft;
+            bytes -= bytesLeft;
+        }
+    }
+
+    if (bytes) {
+        // Allocate storage alongside the operation structure itself
+        unsigned bytesAlloc = std::max(bytes, kMinBacklogBytes);
+        auto membuf = new uint8_t[sizeof(WriteOperation) + bytesAlloc];
+        WriteOperation* op = new (membuf) WriteOperation;
+        conn->fWriteOps.emplace_back(op);
+
+        // TODO: asyncId
+        op->fAllocSize = bytesAlloc;
+        op->fNotify.asyncId = /* TODO */nullptr;
+        op->fNotify.buffer = membuf + sizeof(WriteOperation);
+        op->fNotify.bytes = bytes;
+        op->fNotify.bytesProcessed = 0;

```

Comment:
The NT socket sets `bytesProcessed` to `bytes` instead of 0. Having slightly different record keeping isn't the biggest problem, but looking at this difference is on my list of things to do. Maybe the NT socket has some trick that would simplify record keeping a bit on the transaction.